internal/pagination/pagination2.go

```diff
@@ -0,0 +1,182 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: BUSL-1.1
+
+package pagination
+
+import (
+	"context"
+	"time"
+
+	"github.com/hashicorp/boundary/internal/boundary"
+	"github.com/hashicorp/boundary/internal/errors"
+	"github.com/hashicorp/boundary/internal/refreshtoken"
+)
+
+// ListResponse2 represents the response from the paginated list operation.
+type ListResponse2[T boundary.Resource] struct {
```

Comment:
This type will be renamed to `ListResponse` when we can remove the old pagination logic.